### DISCUSSION TITLE
Comparison to fp-ts?

### DISCUSSION THREAD

**Discussion by xddq** [ramda/ramda#3185] [Q&A] [answered]
I am trying to get my head around functional programming and am not sure with which library I should get started. I guess ramda is bigger and has more docu etc.. so there is that.

Do you happen to know [fp-ts](https://github.com/gcanti/fp-ts)  and could draw a line. Are these similar libraries, when should one use which of them?

thank you in advance

**Comment by CrossEye** [comment, marked as answer]
I absolutely love what I've seen of fp-ts.  But I've never used it because I don't use TypeScript by choice.  So I can't speak to details of features.  Ramda has been around longer and has (I'd bet) a significantly larger user-base, and therefore probably more learning resources.  But if your focus is TypeScript, I'd bet that fp-ts is a great choice.

Some years ago, I wrote [a long answer][so] to a StackOverflow question detailing the relationship between Ramda, FantasyLand, and Folktale, and explained that

> Folktale is a collection of types implementing various parts of the Fantasy-land specification and a small collection of companion utility functions. These types are things like `Maybe`, `Either`, `Task` (very similar to what is elsewhere called a `Future`, and a more lawful cousin to a `Promise`), and `Validation`
> [ ... ]
> Ramda [ ... ] is a very different type of library. It does not provide new types for you. Instead, it provides functions to make it easier to operate on existing types. It is built around the notions of composing smaller functions into larger ones, of working with immutable data, of avoiding side-effects.

`fp-ts` seems to straddle those worlds, providing a number of useful types that match the FantasyLand types as well as a large collection of utility functions (many derived from those type definitions.)  I would actually love to build a pure JS library along the same lines, but that's a very different focus from Ramda's attempt to make FP simpler to do in JS.

@gcanti: Do you want to weigh in here?


  [so]: https://stackoverflow.com/a/33130194

**Reply by xddq** [reply]
Interesting, thank you for the good comparison (also bringing in fantasyland and folktale)!

Since I prefer typescript I guess I'll dive into fp-ts after my exams.

Do you want the library to use fp-ts in js or because you would enjoy creating it?
If it is for using, couldn't you just compile the .ts library to .js and use it? Could even auto build/host with each release in fp-ts I guess.

**Reply by CrossEye** [reply]
> Do you want the library to use fp-ts in js or because you would enjoy creating it?

Probably a bit of both.

From my limited experience with TS, it seems that libraries written for TS end up with very different concerns and very different feels from ones written for JS.  If I wanted a compile-to-JS environment, I would probably be more interested in something like PureScript, or possibly ClojureScript, 

But I also like trying to write these things; the process of figuring it all out is some of my favorite time coding.
